src/Razor/src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.MvcApplicationPartsDiscovery.targets

```diff
@@ -67,7 +67,16 @@ Copyright (c) .NET Foundation. All rights reserved.
       OutputFile="$(_MvcApplicationPartAttributeGeneratedFile)"
       Condition="'@(_ApplicationPartAssemblyNames->Count())' != '0'" />
 
-    <ItemGroup Condition="'@(_ApplicationPartAssemblyNames->Count())' != '0'">
+    <!--
+      If the generated attribute file exists, but no assembly names were discovered,
+      it suggests that the assembly references were changed to no longer point to MVC.
+      In this case, delete the file so that future incremental builds can no longer pick it up.
+    -->
+    <Delete
+      File="$(_MvcApplicationPartAttributeGeneratedFile)"
+      Condition="'@(_ApplicationPartAssemblyNames->Count())' == '0' AND Exists('$(_MvcApplicationPartAttributeGeneratedFile)')" />
+
+    <ItemGroup Condition="Exists('$(_MvcApplicationPartAttributeGeneratedFile)')">
```

Comment:
The issue was that in incremental builds where your app recompiles, but references haven't changed, `_ApplicationPartAssemblyNames` isn't defined since the discovery task does not run.

The fix is to use the presence of the file to determine if it needs to be included. I added a delete in case you went from App references Mvc -> App does not reference Mvc, but writing a test for it is really iffy. I verified it works correctly locally.